lib/CL/clBuildProgram.c

```diff
@@ -293,8 +340,73 @@ CL_API_SUFFIX__VERSION_1_0
                      "not available for the program, or do not exist\n");
 
   program->build_status = CL_BUILD_SUCCESS;
+  /* TODO should not unlock program while adding default kernels */
   POCL_UNLOCK_OBJ(program);
+
+  /* set up all program kernels */
+  /* TODO probably wrong to assume */
+  assert(program->num_kernels == 0);
+  program_set_num_kernels(program);
+  program_set_kernel_names(program);
+  program->default_kernels = calloc(program->num_kernels, sizeof(cl_kernel));
+  for (i=0; i < program->num_kernels; i++)
+    {
+      program->default_kernels[i] =
+          POname(clCreateKernel)(program,
+                                 program->kernel_names[i],
+                                 &error);
+      POCL_GOTO_ERROR_COND((error != CL_SUCCESS),
+                           CL_BUILD_PROGRAM_FAILURE);
+    }
+
+  _cl_command_node cmd;
```

Comment:
Yes, dev->ops->compile() takes a command_node argument which has a cl_kernel. I think it needs more than the cachedir, i can take a look at what precisely is required.